# 异常

Java的基本理念是“结构不佳的代码不能运行”。

发现错误的理想时机是在编译阶段，也就是在你试图运行程序之前。然而，编译期间并不能找出所有的错误，余下的问题必须在运行期间解决。

## 1 异常概述

Throwable是所有异常类型的基类，该类有两个直接子类：

- Error：Error类层次结构描述了Java运行时系统的内部错误和资源耗尽错误。
- Exception
  - RuntimeException：由于程序错误导致的异常属于RuntimeException
  
    派生于RuntimeException的异常包含下面几种情况：
  
    - 错误的类型转换。
    - 数组访问越界。
    - 访问null指针。
  
  - 其他异常：程序本身没有问题，但由于像I/O错误这类问题导致的异常属于其他异常
  
    不是派生于RuntimeException的异常包括：
  
    - 试图在文件尾部后面读取数据。
    - 试图打开一个不存在的文件。
    - 试图根据给定的字符串查找Class对象，而这个字符串表示的类并不存在

![image-20201126204633607](X:\Users\11077\AppData\Roaming\Typora\typora-user-images\image-20201126204633607.png)

## 2 异常的分类

受查异常和非受查异常

Java语言规范将派生于Error或RuntimeException类的所有异常称为非受查（unchecked）异常，所有其他的异常称为受查（checked）异常。

### 2.1 受查异常（非Error或RuntimeException类）

受查异常会在编译时被检测。

> 也可以理解为可控式异常：可以预知的错误，例如从文件中读取数据、对数据库进行操等，在程序编译时就能对程序中可能存在的错误进行处理，并给出具体的错误信息。

如果一个方法中的代码会抛出受查异常，则该方法**必须包含异常处理**，即 try-catch 代码块，或在方法签名中用 throws 关键字**声明**该方法可能会抛出的**受查异常**，**否则编译无法通过**。如果一个方法可能抛出多个受查异常类型，就必须在方法的签名处列出所有的异常类。

案例：此时编译器会报错

```java
private static void readFile(String filePath) {
    File file = new File(filePath);
    String result;
    BufferedReader reader = new BufferedReader(new FileReader(file));
    while ((result = reader.readLine()) != null) {
        System.out.println(result);
    }
    reader.close();
}
```

![image-20201126205726529](https://yeyangshu-picgo.oss-cn-shanghai.aliyuncs.com/img/image-20201126205726529.png)

解决方式一：通过 throws 关键字声明可能抛出的异常

```java
private static void readFile(String filePath) throws IOException {
    File file = new File(filePath);
    String result;
    BufferedReader reader = new BufferedReader(new FileReader(file));
    while ((result = reader.readLine()) != null) {
        System.out.println(result);
    }
    reader.close();
}
```

解决方式一：try-catch 处理异常

```java
private static void readFile(String filePath) {
    File file = new File(filePath);
    String result;
    BufferedReader reader;
    try {
        reader = new BufferedReader(new FileReader(file));
        while ((result = reader.readLine()) != null) {
            System.out.println(result);
        }
        reader.close();
    } catch (IOException e) {
        e.printStackTrace();
    }
}
```

#### 2.1.1 常见的可控式异常

![image-20201126210542824](https://yeyangshu-picgo.oss-cn-shanghai.aliyuncs.com/img/image-20201126210542824.png)

### 2.2 非受查异常

非受查异常不会在编译时被检测。JAVA 中 Error 和 RuntimeException 类的子类属于非受查异常.

> 也可以理解为运行时异常：有些错误不能被编译器检测到，例如，在进行除法运算时，除数为零，Java编译器是检测不到的，因而能正常编译，但是在运行时就会发生异常。

案例：

```java
public class T {
    public static void main(String[] args) {
        Object o = new Integer(0);
        System.out.println((String)o);
    }

    /**
     * Exception in thread "main" java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.String
     */
}
```

#### 2.2.1 常见的运行时异常

![image-20201126210713773](https://yeyangshu-picgo.oss-cn-shanghai.aliyuncs.com/img/image-20201126210713773.png)

## 3 获取异常信息

Java 中的 `java.lang.Throwable`提供了获得异常信息的方法。

![image-20201126210938127](https://yeyangshu-picgo.oss-cn-shanghai.aliyuncs.com/img/image-20201126210938127.png)

## 4 处理异常（try/catch）

要明白异常是如何被捕获的，首先必须理解监控区域（guarded region）的概念。它是一段可能产生异常的代码，并且后面跟着处理这些异常的代码。

### 4.1 try 块

如果在方法内部抛出了异常（或者在方法内部调用的其他方法抛出了异常），这个方法将在抛出异常的过程中结束。要是不希望此方法就此结束，可以在方法内设置一个特殊的块来捕获异常。因为在这个块里“尝试”各种（可能产生异常的）方法调用，所以称为try块。

```java
try {
    // Code that might generate exceptions
}
```

### 4.2 异常处理程序 catch

抛出的异常必须在某处得到处理。这个地点就是**异常处理程序**，而且要针对每个要捕获的异常，得准备相应的处理程序。异常处理程序紧跟在try块之后，以关键字catch表示：

```java
try {
    // Code that might generate exceptions
} catch(Type type) {
    // Handle exceptions of type
}
```

异常处理程序必须紧跟在 try 块之后。当异常被抛出时，异常处理机制将负责搜寻参数与异常类型相匹配的第一个处理程序。然后进入 catch 子句执行，此时认为异常得到了处理。

注意在 try 块的内部，许多不同的方法调用可能会产生类型相同的异常，而你只需要提供一个针对此类型的异常处理程序。

### 4.3 使用 try…catch 处理异常

### 4.4 使用 try…catch…finally 处理异常

### 4.5 使用 try…finally 处理异常

## 5 抛出异常（throws 声明/throw 语句）

### 5.1 使用 throws 声明抛出异常

一个方法不仅要告诉编译器将要返回什么类型的值，还要告诉编程器可能返回的错误。

方法应该在其首部申明所有可能抛出的异常。

Java鼓励人们把方法可能抛出的异常告知此方法的客户端程序员。这是一种优雅的做法，它使得调用者能确切知道写什么样的代码可以捕获所有最潜在的异常。Java提供了相应的预发（并强制使用了这个语法），使得你能以礼貌的方式告知客户端程序员某个方法可能会抛出的异常类型，然后客户端程序员就可以进行响应的处理。这就是**异常说明**。它属于方法声明的一部分，紧跟在形式参数列表之后。

throws 通常用于方发声明，当方法中可能存在异常，却不想在方法中对异常进行处理时，就可以在声明方法时使用 throws 声明抛出异常，然后在调用该方法的其他方法中对异常进行处理（如使用 try…catch 语句或使用 throws 继续抛出异常）。

```java
void f() throws TooBig, TooSmall {
    // ...
}
```

就表示此方法不会抛出任何异常。

### 5.2 使用 throw 语句抛出异常

异常情形（exceptional condition）是指阻止当前方法或作用域继续执行的问题。

使用new在堆上创建异常对象，当前的执行路径被终止，并且从当前环境中弹出对异常对象的引用。此时，异常处理机制接管程序，并开始寻找一个恰当的地方来继续执行程序。这个恰当的地方就是**异常处理程序。**

在通常情况下，程序发生错误时系统会自动抛出异常，而有时希望程序**自行抛出异常**，就可以使用 throw 语句来实现。

```java
if (t == null) {
    // 自行抛出异常
    throw new NullPoninterException();
}
```

抛出了异常，在当前环境中就不必为这个问题操心，它将会在别的地方得到处理。

异常最重要的方面之一就是如果发生问题，它们将不允许程序沿着其正常的路径继续走下去。

## 6 自定义异常

当我们在代码中需要抛出异常时，尽量使用JDK已定义的异常类型。例如，参数检查不合法，应该抛出`IllegalArgumentException`：

```java
static void process1(int age) {
    if (age <= 0) {
        throw new IllegalArgumentException();
    }
}
```

在一个大型项目中，可以自定义新的异常类型，但是，保持一个合理的异常继承体系是非常重要的。

一个常见的做法是自定义一个`BaseException`作为“根异常”，然后，派生出各种业务类型的异常。

`BaseException`需要从一个适合的`Exception`派生，通常建议从`RuntimeException`派生：

```
public class BaseException extends RuntimeException {
}
```

其他业务类型的异常就可以从`BaseException`派生：

```
public class UserNotFoundException extends BaseException {
}

public class LoginFailedException extends BaseException {
}

...
```

自定义的`BaseException`应该提供多个构造方法：

```
public class BaseException extends RuntimeException {
    public BaseException() {
        super();
    }

    public BaseException(String message, Throwable cause) {
        super(message, cause);
    }

    public BaseException(String message) {
        super(message);
    }

    public BaseException(Throwable cause) {
        super(cause);
    }
}
```

## 7 异常使用原理

使用异常的时机

- 异常处理不能代替简单的测试

  只有在异常的时候使用异常机制

- 不要过分的细化异常

  目标：将正常处理与错误处理分开

- 利用异常层次结构

  不要只抛出

- 不要压制异常

## 8 经典样例








